src/array/cpu/id_hash_map.cc

```diff
@@ -0,0 +1,198 @@
+inline size_t GetMapSize(size_t num) {
+  size_t capacity = 1;
+  return capacity << static_cast<size_t>(1 + std::log2(num * 3));
+}
+}  // namespace
+
+namespace dgl {
+namespace aten {
+
+template <typename IdType>
+IdType IdHashMap<IdType>::CompareAndSwap(
+    IdType* ptr, IdType old_val, IdType new_val) {
+#ifdef _MSC_VER
+  if (sizeof(IdType) == 4) {
+    return _InterlockedCompareExchange(
+        reinterpret_cast<LONG*>(ptr), new_val, old_val);
+  } else if (sizeof(IdType) == 8) {
+    return _InterlockedCompareExchange64(
+        reinterpret_cast<LONGLONG*>(ptr), new_val, old_val);
+  } else {
+    LOG(FATAL) << "ID can only be int32 or int64";
+  }
+#elif __GNUC__  // _MSC_VER
+  return __sync_val_compare_and_swap(ptr, old_val, new_val);
+#else           // _MSC_VER
+#error "CompareAndSwap is not supported on this platform."
+#endif  // _MSC_VER
+}
+
+template <typename IdType>
+IdHashMap<IdType>::IdHashMap() : mask_(0) {
+  // Used to deallocate the memory in hash_map_ with device api
+  // when the pointer is freed.
+  auto deleter = [](Mapping* mappings) {
+    if (mappings != nullptr) {
+      DGLContext ctx = DGLContext{kDGLCPU, 0};
+      auto device = DeviceAPI::Get(ctx);
+      device->FreeWorkspace(ctx, mappings);
+    }
+  };
+  hash_map_ = {nullptr, deleter};
+}
+
+template <typename IdType>
+IdArray IdHashMap<IdType>::Init(const IdArray ids) {
+  CHECK_EQ(ids.defined(), true);
+  const IdType* ids_data = ids.Ptr<IdType>();
+  const size_t num_ids = static_cast<size_t>(ids->shape[0]);
+  // Make sure `ids` is not 0 dim.
+  CHECK_GT(num_ids, 0);
+  size_t capacity = GetMapSize(num_ids);
+  mask_ = static_cast<IdType>(capacity - 1);
+
+  auto ctx = DGLContext{kDGLCPU, 0};
+  hash_map_.reset(static_cast<Mapping*>(
+      DeviceAPI::Get(ctx)->AllocWorkspace(ctx, sizeof(Mapping) * capacity)));
+  memset(hash_map_.get(), -1, sizeof(Mapping) * capacity);
+
+  // This code block is to fill the ids into hash_map_.
+  IdArray unique_ids = NewIdArray(num_ids, ctx, sizeof(IdType) * 8);
+  // Use `int16_t` instead of `bool`. As vector<bool> is an exception
+  // for whom updating different elements from different threads is unsafe.
+  // see https://en.cppreference.com/w/cpp/container#Thread_safety.
+  std::vector<int16_t> valid(num_ids);
+  auto thread_num = compute_num_threads(0, num_ids, kGrainSize);
+  std::vector<size_t> block_offset(thread_num + 1, 0);
+  IdType* unique_ids_data = unique_ids.Ptr<IdType>();
+
+  // Insert all elements in this loop.
+  parallel_for(0, num_ids, kGrainSize, [&](int64_t s, int64_t e) {
+    size_t count = 0;
+    for (int64_t i = s; i < e; i++) {
+      Insert(ids_data[i], &valid, i);
+      count += valid[i];
+    }
+    block_offset[omp_get_thread_num() + 1] = count;
+  });
+
+  // Get ExclusiveSum of each block.
+  std::partial_sum(
+      block_offset.begin() + 1, block_offset.end(), block_offset.begin() + 1);
+  unique_ids->shape[0] = block_offset.back();
```

Comment:
Is assigning shape values OK to do?  I feel it's very dangerous.  It's safer to compute the shape first and then allocate the `unique_ids` array.

Reply:
It looks safe because I find some other usages like this and the memory will not leaked. We can not allocate ahead because it is known after the data is filled in.